src/node.rs

```diff
@@ -285,10 +288,21 @@ impl Node {
     }
 
     /// Remove messages that have already been sent from `messages`, and update the filter.
-    fn filter_messages(&mut self, mut messages: Vec<Message>) -> Vec<Message> {
-        messages.retain(|msg| !self.message_filter.contains(msg));
-        self.message_filter.extend(messages.clone());
-        messages
+    fn filter_messages(&mut self, messages: Vec<Message>) -> Vec<Message> {
+        let mut filtered = vec![];
+        for message in messages {
+            let mut hasher = DefaultHasher::new();
+            message.hash(&mut hasher);
+            let hash = hasher.finish();
+            if !self.message_filter.contains(&hash) {
```

Comment:
Yes, I agree, profiling doesn't show message filtering to be a major drain on cycles, so it's just the mem usage we should aim to reduce...